From 44658286e3ce1d23f74f520a69628e916144c7f6 Mon Sep 17 00:00:00 2001
From: Fugang Duan <b38611@freescale.com>
Date: Thu, 26 Mar 2015 15:26:18 +0800
Subject: [PATCH] MLK-10472-1 gpio: max7322: modify the max7322 driver to
 support device tree interface.

The max7322 driver can only support platform data interface. This patch
modify drvier to also support device tree interface.

(cherry-picked from commit: e0954ca10fbce4a820e7882b3f87cb78907fd987)

Signed-off-by: Luwei Zhou <b45643@freescale.com>
Signed-off-by: Fugang Duan <B38611@freescale.com>
---
 drivers/gpio/gpio-max732x.c | 23 +++++++++++++++--------
 1 file changed, 15 insertions(+), 8 deletions(-)

diff --git a/drivers/gpio/gpio-max732x.c b/drivers/gpio/gpio-max732x.c
index 36cb290..d607a1a 100644
--- a/drivers/gpio/gpio-max732x.c
+++ b/drivers/gpio/gpio-max732x.c
@@ -21,6 +21,7 @@
 #include <linux/irq.h>
 #include <linux/i2c.h>
 #include <linux/i2c/max732x.h>
+#include <linux/reset.h>
 
 
 /*
@@ -457,7 +458,7 @@ static int max732x_irq_setup(struct max732x_chip *chip,
 	int has_irq = max732x_features[id->driver_data] >> 32;
 	int ret;
 
-	if (pdata->irq_base && has_irq != INT_NONE) {
+	if (pdata && pdata->irq_base && has_irq != INT_NONE) {
 		int lvl;
 
 		chip->irq_base = pdata->irq_base;
@@ -515,7 +516,7 @@ static int max732x_irq_setup(struct max732x_chip *chip,
 	struct max732x_platform_data *pdata = dev_get_platdata(&client->dev);
 	int has_irq = max732x_features[id->driver_data] >> 32;
 
-	if (pdata->irq_base && has_irq != INT_NONE)
+	if (pdata && pdata->irq_base && has_irq != INT_NONE)
 		dev_warn(&client->dev, "interrupt support not compiled in\n");
 
 	return 0;
@@ -570,6 +571,7 @@ static int max732x_setup_gpio(struct max732x_chip *chip,
 	gc->ngpio = port;
 	gc->label = chip->client->name;
 	gc->owner = THIS_MODULE;
+	gc->dev = &chip->client->dev;
 
 	return port;
 }
@@ -582,12 +584,13 @@ static int max732x_probe(struct i2c_client *client,
 	struct i2c_client *c;
 	uint16_t addr_a, addr_b;
 	int ret, nr_port;
+	int gpio_base = -1;
 
 	pdata = dev_get_platdata(&client->dev);
-	if (pdata == NULL) {
+	if (pdata == NULL)
 		dev_dbg(&client->dev, "no platform data\n");
-		return -EINVAL;
-	}
+	else
+		gpio_base = pdata->gpio_base;
 
 	chip = devm_kzalloc(&client->dev, sizeof(struct max732x_chip),
 			GFP_KERNEL);
@@ -595,7 +598,11 @@ static int max732x_probe(struct i2c_client *client,
 		return -ENOMEM;
 	chip->client = client;
 
-	nr_port = max732x_setup_gpio(chip, id, pdata->gpio_base);
+	ret = device_reset(&client->dev);
+	if (ret == -ENODEV)
+		return -EPROBE_DEFER;
+
+	nr_port = max732x_setup_gpio(chip, id, gpio_base);
 
 	addr_a = (client->addr & 0x0f) | 0x60;
 	addr_b = (client->addr & 0x0f) | 0x50;
@@ -636,7 +643,7 @@ static int max732x_probe(struct i2c_client *client,
 	if (ret)
 		goto out_failed;
 
-	if (pdata->setup) {
+	if (pdata && pdata->setup) {
 		ret = pdata->setup(client, chip->gpio_chip.base,
 				chip->gpio_chip.ngpio, pdata->context);
 		if (ret < 0)
@@ -657,7 +664,7 @@ static int max732x_remove(struct i2c_client *client)
 	struct max732x_chip *chip = i2c_get_clientdata(client);
 	int ret;
 
-	if (pdata->teardown) {
+	if (pdata && pdata->teardown) {
 		ret = pdata->teardown(client, chip->gpio_chip.base,
 				chip->gpio_chip.ngpio, pdata->context);
 		if (ret < 0) {
-- 
1.8.0

